File: 3900_longest_balanced_substring.py
from collections import Counter, defaultdict, OrderedDict
from hashlib import md5

##  "Don't optimise until you've measured." - David Patterson, Computer Organisation

##  Wrote a coherent BFS search, but potentially only a DFS search works for test 963/992.
##  Memory limit exceeded (perhaps I should've seen it coming).

class Solution:

    def longestBalanced(self, s: str) -> int:

        ##print("s:", s)

        if '1' not in s or '0' not in s:
            return 0

        start = time.time()

        bal_string_chks = {}

        ##  Don't exactly need a key-value store, but do want O(1) lookups.

        ss_counts = OrderedDict()

        biggest_str = True

        ##  Generate all substrings in useful order (as opposed to s[i:j]).
        ##  As in, largest string to smallest string.

        if len(s) % 2 == 0:
            init_q = len(s)
        else:
            init_q = len(s) - 1

        ##print("init_q:", init_q)
        
        w_quantity = [0, 0]
        w_quantity_batch = 0

        for q in range(init_q, 0, -2):

            print(w_quantity)

            w_quantity = [w_quantity_batch, w_quantity[0]]
            w_quantity_batch = 0

            while (w_quantity[1] > 0):
                ss_counts.popitem()
                w_quantity[1] -= 1

            running_counter = {'0': 0, '1': 0}
            
            print("window size:", q)

            for i in range(0, len(s)):

                print("ss counts len:", len(ss_counts))

                # print("i:", i)

                ##  Extract substring.

                sub_s = s[i:i + q]

                ##  Simplest sliding window implementation involves a break.

                if len(sub_s) < q:
                    break
                else:
                    w_quantity_batch += 1

                ##print(sub_s)

                ##  Only call Counter() once, then re-use results.

                if biggest_str == True:
                    biggest_str = False
                    ss_counts[self.hsh(sub_s)] = running_counter | dict(Counter(sub_s))

                    ##print(ss_counts)

                if sub_s not in ss_counts:

                    ##  Count 1s and 0s using minimal computation.
                    running_counter, ss_counts = self.calculate_counts(s, sub_s, ss_counts, dict(running_counter), i, q)

                else:
                    ##  Else simply lookup.
                    running_counter = dict(ss_counts[self.hsh(sub_s)])

                ##print("ss_counts:", ss_counts)

                if sub_s not in bal_string_chks:
                    bal_string_chks[sub_s] = self.check_string(s, sub_s, ss_counts[self.hsh(sub_s)], i)

                if bal_string_chks[sub_s] > -1:

                    return bal_string_chks[sub_s]
            
        return 0


    def calculate_counts(self, s: str, sub_s: str, ss_counts: dict[str, int], running_counter: dict[str, int], i: int, q: int) -> tuple[dict[str, int], dict[str, int]]:

        ##  Counter() calls are the biggest bottleneck in the program,
        ##    so this part becomes a lengthy caching workaround.

        ##  If this is a new sliding window size, use count of larger string.

        ##print("sub_s:", sub_s)

        to_prepend, to_append = '', ''
        if i != 0:
            to_prepend = s[i - 2:i]
        if to_prepend == '':
            to_prepend = '__'

        ##print("i + q:", i+q)

        if i + q < len(s):
            to_append = s[i + q:i + q + 2]
        if to_append == '':
            to_append = '__'

        extra_chars = (to_prepend, to_append)

        if running_counter == {'0': 0, '1': 0 }:

            ##print("ss_counts:", ss_counts)

            ##print("extra_chars:", extra_chars)

            ##  Algorithm will look ahead of substring, before behind.

            if self.hsh(f"{extra_chars[0]}{sub_s}") in ss_counts:
                running_counter = dict(ss_counts[self.hsh(f"{extra_chars[0]}{sub_s}")])
                running_counter[extra_chars[0][0]] -= 1
                running_counter[extra_chars[0][1]] -= 1
                ##print("Updating running counter via prepended substring")
                ##print(sub_s, extra_chars, running_counter)
                
            elif self.hsh(f"{sub_s}{extra_chars[1]}") in ss_counts:
                running_counter = dict(ss_counts[self.hsh(f"{sub_s}{extra_chars[1]}")])
                running_counter[extra_chars[1][0]] -= 1
                running_counter[extra_chars[1][1]] -= 1
                ##print("Updating running counter via appended substring")
                ##print(sub_s, extra_chars, running_counter)
            
            else:
                ##  If biggest_str is an odd length the above won't be reached.
                ##  Simplest to call Counter() once more.
                ##print("Running Counter()")
                ss_counts[self.hsh(sub_s)] = Counter(sub_s)


        ##  Else adjust running_counter.
        else:

            ##print("Updating running counter")

            running_counter[s[i-1]] -= 1
            running_counter[sub_s[-1]] += 1

            ##print(sub_s, extra_chars, running_counter)

        ##  Store for potential future lookups.
        ss_counts[self.hsh(sub_s)] = dict(running_counter)

        ##print(ss_counts)

        return running_counter, ss_counts


    def check_string(self, s: str, sub_s: str, counts: dict[str, int], pos: int) -> int:

        target = None

        ##  If equal quantities of 1 and 0, substring is best.

        ##print("counts", counts)

        if counts['0'] == counts['1']:
            return len(sub_s)

        ##  If there is only 1 character difference: viable candidate.

        elif counts['0'] - 1 == counts['1'] + 1:

            target = '1'

        elif counts['0'] + 1 == counts['1'] - 1:

            target = '0'

        ##  Compare with rest of string, seeking a 1 or a 0.

        if target is not None:

            pre_str = s[0:pos]

            post_str = s[(pos + len(sub_s)):]

            if target in pre_str or target in post_str:

                return len(sub_s)

        return -1

    def hsh(self, s: str) -> str:

        ##  Reducing memory consumption of long (1, 0) strings, via hash.
        ##  In exchange for CPU cycles, that is.
        ##  The cutting of md5 

        return s

        return adler32(str.encode(s))

        return md5(s.encode('utf-8')).hexdigest()
    

File: 3900_longest_balanced_substring.go
package main

import (
	"strings"
)

//  The following is an automated port from the Python version, 3900_longest_balanced_substring.py, via Gemini 3.6 Flash LLM.
//  Only 2 lines were changed, in very small ways, in calculateCounts(), and comments added.

//  955/992 test pass, failure due to execution timeout,
//    i.e. worse performance than the Python version (but not much in it).


type counts map[byte]int

func copyCounts(c counts) counts {
	newC := make(counts, len(c))
	for k, v := range c {
		newC[k] = v
	}
	return newC
}

// orderedMap string -> counts
type orderedCountsMap struct {
	m    map[string]counts
	keys []string
}

func newOrderedCountsMap() *orderedCountsMap {
	return &orderedCountsMap{
		m:    make(map[string]counts),
		keys: make([]string, 0),
	}
}

func (o *orderedCountsMap) get(key string) (counts, bool) {
	val, ok := o.m[key]
	return val, ok
}

func (o *orderedCountsMap) set(key string, val counts) {
	if _, exists := o.m[key]; !exists {
		o.keys = append(o.keys, key)
	}
	o.m[key] = val
}

func (o *orderedCountsMap) popFirst() {
	if len(o.keys) == 0 {
		return
	}
	firstKey := o.keys[0]
	o.keys = o.keys[1:]
	delete(o.m, firstKey)
}

// orderedMap string -> int
type orderedIntMap struct {
	m    map[string]int
	keys []string
}

func newOrderedIntMap() *orderedIntMap {
	return &orderedIntMap{
		m:    make(map[string]int),
		keys: make([]string, 0),
	}
}

func (o *orderedIntMap) get(key string) (int, bool) {
	val, ok := o.m[key]
	return val, ok
}

func (o *orderedIntMap) set(key string, val int) {
	if _, exists := o.m[key]; !exists {
		o.keys = append(o.keys, key)
	}
	o.m[key] = val
}

func (o *orderedIntMap) popFirst() {
	if len(o.keys) == 0 {
		return
	}
	firstKey := o.keys[0]
	o.keys = o.keys[1:]
	delete(o.m, firstKey)
}

func hsh(s string) string {
	return s
}

func makeCounter(s string) counts {
	c := counts{'0': 0, '1': 0}
	for i := 0; i < len(s); i++ {
		c[s[i]]++
	}
	return c
}

func checkString(s, subS string, c counts, pos int) int {
	var target byte = 0

	if c['0'] == c['1'] {
		return len(subS)
	} else if c['0']-1 == c['1']+1 {
		target = '1'
	} else if c['0']+1 == c['1']-1 {
		target = '0'
	}

	if target != 0 {
		preStr := s[:pos]
		postStr := s[pos+len(subS):]

		if strings.IndexByte(preStr, target) != -1 || strings.IndexByte(postStr, target) != -1 {
			return len(subS)
		}
	}

	return -1
}

func calculateCounts(s, subS string, ssCounts *orderedCountsMap, runningCounter counts, i, q int) (counts, *orderedCountsMap) {
	toPrepend := ""
	toAppend := ""

  //  FOLLOWING LINE MANUALLY EDITED FROM AUTOMATED PORT.
	if i > 1 {
		toPrepend = s[i-2 : i]
	}
	if toPrepend == "" {
		toPrepend = "__"
	}

  //  FOLLOWING LINE MANUALLY EDITED FROM AUTOMATED PORT.
	if i+q+2 <= len(s) {
		toAppend = s[i+q : i+q+2]
	}
	if toAppend == "" {
		toAppend = "__"
	}

	extraChars := [2]string{toPrepend, toAppend}

	if runningCounter['0'] == 0 && runningCounter['1'] == 0 {
		key1 := hsh(extraChars[0] + subS)
		key2 := hsh(subS + extraChars[1])

		if val, exists := ssCounts.get(key1); exists {
			runningCounter = copyCounts(val)
			runningCounter[extraChars[0][0]]--
			runningCounter[extraChars[0][1]]--
		} else if val, exists := ssCounts.get(key2); exists {
			runningCounter = copyCounts(val)
			runningCounter[extraChars[1][0]]--
			runningCounter[extraChars[1][1]]--
		} else {
			ssCounts.set(hsh(subS), makeCounter(subS))
		}
	} else {
		runningCounter[s[i-1]]--
		runningCounter[subS[len(subS)-1]]++
	}

	ssCounts.set(hsh(subS), copyCounts(runningCounter))
	return runningCounter, ssCounts
}

func longestBalanced(s string) int {
	if !strings.Contains(s, "1") || !strings.Contains(s, "0") {
		return 0
	}

	balStringChks := newOrderedIntMap()
	ssCounts := newOrderedCountsMap()
	biggestStr := true

	initQ := len(s)
	if initQ%2 != 0 {
		initQ = len(s) - 1
	}

	wQuantity := [2]int{0, 0}
	wQuantityBatch := 0

	for q := initQ; q > 0; q -= 2 {
		wQuantity = [2]int{wQuantityBatch, wQuantity[0]}
		wQuantityBatch = 0

		for wQuantity[1] > 0 {
			ssCounts.popFirst()
			balStringChks.popFirst()
			wQuantity[1]--
		}

		runningCounter := counts{'0': 0, '1': 0}

		for i := 0; i < len(s); i++ {
			if i+q > len(s) {
				break
			}

			subS := s[i : i+q]

			if biggestStr {
				biggestStr = false
				cnts := makeCounter(subS)
				rc := copyCounts(runningCounter)
				for k, v := range cnts {
					rc[k] = v
				}
				ssCounts.set(hsh(subS), rc)
			}

			if _, exists := ssCounts.get(hsh(subS)); !exists {
				runningCounter, ssCounts = calculateCounts(s, subS, ssCounts, copyCounts(runningCounter), i, q)
				wQuantityBatch++
			} else {
				val, _ := ssCounts.get(hsh(subS))
				runningCounter = copyCounts(val)
			}

			if _, exists := balStringChks.get(subS); !exists {
				cntVal, _ := ssCounts.get(hsh(subS))
				balStringChks.set(hsh(subS), checkString(s, subS, cntVal, i))
			}

			if res, _ := balStringChks.get(hsh(subS)); res > -1 {
				return res
			}
		}
	}

	return 0
}
